### PR TITLE
rpc: allow to pass RPC login via RPC_LOGIN env var

### PR DESCRIPTION
**Pull request by ph4r05**
- passing by parameter is insecure as it is shown in the process list
- using environment variable for passing the password is more secure with this respect

Thanks for considering.

### CODE REVIEW

**Comment by moneromooo-monero** [comment]
Do you mean "it can be seen in ps" ?

**Comment by ph4r05** [comment]
Correct

**Comment by ph4r05** [comment]
It may depend on the OS but I mean, the password in the argument is leaked a) to the process table, visible to all other non-root users running `ps` b) to the shell history, while env var should be inaccessible for all non-root users.

**Comment by ph4r05** [comment]
I've just checked and passing the password via stdin works also well - another reasonably secure option.

**Comment by artyomsol** [comment]
As soon as the wallet RPC uses a _digest authentication_ scheme with MD5 hashing would it be convenient and more secure to pass `MD5(<username>:monero-rpc:<password>)` (see [RFC-2617](https://tools.ietf.org/html/rfc2617#section-3.2.2.2)) as a parameter instead of a plain text values.

**Comment by jtgrassie** [comment]
Will this work with both daemon and monero-wallet-rpc?

**Comment by artyomsol** [comment]
@jtgrassie Is it a question to me or topic starter ?

**Comment by jtgrassie** [comment]
It's to @ph4r05, the author of this PR.

**Comment by ph4r05** [comment]
Ah I wasn’t sure either. This PR patches both wallet rpc server and daemon,
the ‘--rpc-login’ argument can be substituted by RPC_LOGIN env var, however
the argument has higher priority.
pá 31. 8. 2018 v 12:08 odesílatel Jethro Grassie <notifications@github.com>
napsal:

**Comment by moneromooo-monero** [comment]
tbh I'm not sure adding another not quite as insecure way is good if you keep the worse way anyway. Why not use the secure way in the firt place ?

**Comment by ph4r05** [comment]
Do you suggest to remove the argument? I found it useful during testing, it
is easy to use, byt I would not use it in the production. Many
daemons/clients offer the same way, both argument, envvar and possibly
stdin.
pá 31. 8. 2018 v 12:40 odesílatel moneromooo-monero <
notifications@github.com> napsal:

**Comment by moneromooo-monero** [comment]
It is useful for testing. File's useful for security. I don't really have anything against adding a middle ground way, but it seems not so useful. Maybe for testing, but the claim is that it is for security, which seems odd. Am I missing something maybe ?

**Comment by moneromooo-monero** [comment]
Hmm, looks like there is no way to pass this via a file actually. I thought there was. So your patch does improve somewat. It'd be better to add file instead though.

**Comment by ph4r05** [comment]
File would be nice too, but I think in some setups the ENV var is quite a
good solution. It is hidden from all non-root users, does not require FS.

I am starting RPC from a different process. Passing by argument is
insecure, files are pain to manage (creation, permissions). Env var is then
simple and effective. I am maybe missing something but I dont see how file
is more secure than env var for non root attacker here.

pá 31. 8. 2018 v 13:15 odesílatel moneromooo-monero <
notifications@github.com> napsal:

File: src/rpc/rpc_args.cpp [hunk 1]
```diff
@@ -82,13 +82,19 @@ namespace cryptonote
...
-      config.login = tools::login::parse(command_line::get_arg(vm, arg.rpc_login), true, [](bool verify) {
-        return tools::password_container::prompt(verify, "RPC server password");
-      });
-      if (!config.login)
+      if (!config.login) {
+        config.login = tools::login::parse(
+            has_rpc_arg ? command_line::get_arg(vm, arg.rpc_login) : std::string(env_rpc_login), true, [](bool verify) {
+              return tools::password_container::prompt(verify, "RPC server password");
+            });
         return boost::none;
```
[the remark this answers is not in the document]
Comment (moneromooo-monero) [reply src/rpc/rpc_args.cpp]: I meant more the return none, and the config.login test above. I'd say the lines should be swapped to make sense to me.

[the remark this answers is not in the document]
Comment (ph4r05) [reply src/rpc/rpc_args.cpp]: Aah true :D it should have be
```
if (!config.login)
        return boost::none;
```

**Pull request merged**